migrations/20200422144952_structure_project_version_continuation_data.js

```diff
@@ -0,0 +1,164 @@
+const moment = require('moment');
+const { get } = require('lodash');
+const { Value } = require('slate');
+const csv = require('csv-stringify');
+
+const LICENCE_NUMBER = /((7|3)0(0|\/|\-)[0-9]{4})|P[0-9A-Z]{8}/g;
+const EXPIRYLONG = /([0-9]{1,2})(st|nd|rd|th)? (Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Sept|Oct|Nov|Dec)(r?uary|ch|il|e|y|ust|tember|ober|ember)?  ?((20)?(19|20|21))/;
+const EXPIRYLONG2 = /(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)(r?uary|ch|ril|e|y|ust|tember|ober|ember)? ([0-9]{1,2})(st|nd|rd|th)? ((20)?(19|20|21))/;
+const EXPIRYSHORT = /([0-9]{1,2})(\/|\-|\.|\s)([0-9]{1,2})(\/|\-|\.|\s)(20(19|20|21))/;
+
+const TYPES = {
+  LICENCE_NUMBER: 'Could not parse licence number',
+  DATE: 'Could not parse expiry date',
+  NEITHER: 'Could not parse version',
+  MULTIPLE: 'Multple licence numbers found'
+};
+
+const transform = (data, versionId, writeCsvLine) => {
+
+  if (!data) {
+    return;
+  }
+  const continuationQuestion = data['expiring-yes'];
+  // not a continuation
+  if (!continuationQuestion) {
+    return;
+  }
+
+  // already patched
+  if (data['project-continuation']) {
+    return;
+  }
+
+  const obj = Value.fromJSON(JSON.parse(continuationQuestion));
+  const text = obj.document.nodes.map(node => node.text.trim()).filter(node => node).join('\n\n');
+
+  const ppl = text.match(LICENCE_NUMBER) || [];
+
+  let date;
+  if (text.match(EXPIRYLONG)) {
+    const matched = text.match(EXPIRYLONG);
+    const day = matched[1];
+    const month = matched[3];
+    const year = matched[7];
+    date = moment(`${day} ${month} ${year}`, 'DD MMM YY');
+  } else if (text.match(EXPIRYLONG2)) {
+    const matched = text.match(EXPIRYLONG2);
+    const day = matched[25];
+    const month = matched[1];
+    const year = matched[29];
+    date = moment(`${day} ${month} ${year}`, 'DD MMM YY');
+  } else if (text.match(EXPIRYSHORT)) {
+    const matched = text.match(EXPIRYSHORT);
+    let day = matched[1];
+    let month = matched[3];
+    const year = matched[6];
+    if (day.length === 1) {
+      day = '0' + day;
+    }
+    if (month.length === 1) {
+      month = '0' + month;
+    }
+    date = moment(`${day} ${month} ${year}`, 'DD MM YY');
+  }
+
+  if (!ppl.length && !date) {
+    console.log(`Cannot parse versionId: ${versionId}`);
+    writeCsvLine && writeCsvLine(TYPES.NEITHER);
+    return;
+  }
+
+  if (!ppl.length) {
+    console.log(`Unable to parse licence number, versionId: ${versionId}`);
+    writeCsvLine && writeCsvLine(TYPES.LICENCE_NUMBER);
+  }
+
+  if (ppl.length > 1) {
+    console.log(`Multple licence numbers found for versionId: ${versionId}`);
+    writeCsvLine && writeCsvLine(TYPES.MULTIPLE);
+  }
+
+  if (!date) {
+    console.log(`Unable to parse expiry date, versionId: ${versionId}`);
+    writeCsvLine && writeCsvLine(TYPES.DATE);
+  }
```

Comment:
Each of these three conditions should return, no?

Reply:
Can we also add a `writeCsvLine` call for "everything is fine"?

Reply:
And include the input text in the csv?

Reply:
I don't think so - if it can't find the expiry it still writes a single item with the licence number, and vice versa